opal/corelib/string.rb

```diff
@@ -2413,12 +2447,167 @@ def freeze
+    // ensure value is a primitive string
+    this.value = value ? value.toString() : '';
+  }
+
+  // We do not simply assign the JavaScript String prototype to MutableString,
+  // instead we create a new prototype Object based on the String prototype.
+  let proto = Object.create(String.prototype);
+  MutableString.prototype = proto;
+
+  // This is important to be able to overwrite property accessors like String.length.
+  // Otherwise, for example by simply subclassing String, length will be a protected
+  // property, which cannot be overwritten. .length must be overwritten of course,
+  // because when the string is mutated its length may change.
+  Object.defineProperty(proto, 'length', { get: function () { return this.value.length; }});
+
+  // Also .toPrimitive, .valueOf and .toString are essential. The JavaScript String
+  // functions have to work on object strings and primitive strings. So that one function
+  // can work on both variants it calls this.valueOf to get the primitive. This is why
+  // the String functions work flawlessly here for MutableString, simply by providing
+  // .valueOf on the prototype.
+  // Also various JavaScript operators and other things use .toPrimitive, .valueOf and
+  // .toString internally.
+  proto[Symbol.toPrimitive] = function (arg) { return this.value; }
+  proto.valueOf = function () { return this.value; }
+  proto.toString = function (arg) { return this.value; }
+  
+  // Opal.bridge will set the prototype of the brdged class to Opal.Object. However
+  // doing so on MutableString would remeove the existing, essential String prototype
+  // copy and leave MutableString non functional at all. So hint Opal.bridge not to
+  // touch the existing prototype.
+  proto.$$bridge_do_not_touch = true;
+
+  // Here now all the functions that mutate this.value.
+  // Though these functions here are simple and do not validate args or anything, this
+  // is done by the Ruby methods above instead, as they need to raise Ruby exceptions
+  // anyway. So usual JavaScript rules apply. 
+  proto.append = function(s) { return this.value += s.toString(); }
+
+  // Make MutableString globally available.
+  Opal.global.MutableString = MutableString;
+
+  // Delete the existing Opal.String so that we can register the new one below.
+  delete Opal.Object.$$const.String;
+  delete Opal.String;
+}
+
+# Essentially we have two String implementations floating around, with the
+# immutable JavaScript String being hidden, mostly, unless its directly accessed
+# as `String`. A "implementation detail".
+# We must take care to keep them both synchronized. Also its important to note,
+# that the Opal.String <-> MutableString bridge is not a "clean" bridge, but one
+# used with some underlying "magic". So even internally within Opal its preferrable
+# to use Opal.String.$new() over new MutableString(). Best practice is not to use
+# MutableString directly at all, except for debugging.
+class String < `MutableString`
```

Comment:
This seems pluggable. Can we perhaps do `corelib/string/mutable.rb` ?

Reply:
Probably, i will experiment further.